### ISSUE TITLE
TPC-H load generator source duplicates data upon `environmentd` restart when `TICK INTERVAL` is specified

### ISSUE THREAD

**Issue by vmarcos** [MaterializeInc/materialize#16259]
### What version of Materialize are you using?

v0.33.0-dev (d9bda8a97)

### How did you install Materialize?

Built from source

### What is the issue?

Starting from `bin/environmentd --release --reset` locally, issue the following commands:

```
CREATE DATABASE tpch;
CREATE SOURCE lgtpch FROM LOAD GENERATOR TPCH (SCALE FACTOR 0.1, TICK INTERVAL '0.1s') FOR ALL TABLES WITH (SIZE ='4');
```

At this point, we can observe the number of rows in of one of the relations, `region`, that does not get affected by streaming updates:

```
tpch=> SELECT count(*) FROM region;
 count 
-------
     5
(1 row)
```

Now, restart the system with `bin/environmentd --release`. We can observe the number of rows in `region` again:

```
tpch=> SELECT count(*) FROM region;
 count 
-------
    10
(1 row)
```

These sizes are twice as large as expected. Another restart of `environmentd` will add the rows again, making the TPC-H relations grow one more time.

Note that it is important to specify `TICK INTERVAL` above. If the generator is created without a `TICK INTERVAL` specification, then the problem does not manifest.

Reference to relevant thread in [Slack](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1669239387792779?thread_ts=1669220818.522629&cid=C01CFKM1QRF).

**Comment by guswynn** [comment]
does this still repro on main?

**Comment by vmarcos** [comment]
Yes, got the same results following the instructions above in v0.33.0-dev (8c2e9a4be). An important detail in the repro instructions is that the restart of `environmentd` must be without `--reset`.

**Comment by chaas** [comment]
@guswynn Are you looking into this/do you know if the storage team is planning to work on this?

**Comment by guswynn** [comment]
@chaas I was not able to repro it locally, even with @vmarcos's exact instructions, so id probably need to work with them directly to try to figure out what is going on!

**Comment by mjibson** [comment]
I'll try this myself too to see if I can repro or not.

**Comment by mjibson** [comment]
This repros for me using exactly the above sql. Using a `size` more than 1 is required to trigger it. Adding dbg to the `offset.offset` part of generator.rs prints:

```
[src/storage/src/source/generator.rs:110] offset.offset = 0
[src/storage/src/source/generator.rs:110] offset.offset = 0
[src/storage/src/source/generator.rs:110] offset.offset = 70
[src/storage/src/source/generator.rs:110] offset.offset = 0
```

so clearly the problem is the various, uh, whatevers aren't doing the correct thing if they aren't the main whatever?

**Comment by mjibson** [comment]
Smallest repro:

```
CREATE SOURCE lgtpch FROM LOAD GENERATOR TPCH (SCALE FACTOR 0, TICK INTERVAL '0.1s') FOR ALL TABLES WITH (SIZE ='2');
```

**Comment by mjibson** [comment]
I also now think I'm wrong about offset being the problem, all the 0 offsets are not active readers and get immediately shut down.
